Rework of x86 intrinsics support

andreittr opened the pull request:
Massive rework of the x86 intrinsics library, featuring:
- update baseline to upstream LLVM 14
- proper support of all x86 intrinsics (depending on build flags), not only a minimal subset
- delegate `-march=` compile flags to main build system; do not interfere
- split off llvm & gcc trees; keeping a single compatible set of files turned out unmaintainable
- gcc headers based off GCC 13
- compatibility patches for other clang & gcc versions
- tested w/ clang 11, 12, 13, 14, 15, 16; gcc 11, 12, 13

andreittr committed:
Remove obsolete metadata files

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
build: Remove CPU feature config options

This change removes the Kconfig options selecting CPU feature set (i.e.,
`-march=` compiler flag) from the configuration of this library.
CPU feature flags are set elsewhere in the build configuration and the
intrinsics library can already seamlessly pick up these features via
preprocessor defines.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
Remove GCC compatibility adaptations

This change removes code changes meant for compatibility with GCC,
reverting the headers to ones released with LLVM 7.0.1.
This is in anticipation of an update of upstream code, as well as a
rework of compiler compatibility.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
Update to LLVM 14.0.6

In addition, this change greatly expands the available feature headers.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
Add compatibility with clang >= 15

Implementations based off headers in LLVM 15.0.0.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
Add compatibility with clang >= 16

Implementations based off headers in LLVM 16.0.0.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
Add compatibility with clang <= 13

Implementations based off headers in LLVM 13.0.0.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
build: Make clang compatibility explicit

This change makes explicit that this library only supports clang by
renaming the include directory and only adding it to the include path if
compiling with clang.
A future change may thus explicitly add GCC-compatible headers.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
Implement support for GCC

This change adds on-par support for GCC by providing the native x86
intrinsics headers from gcc release 13.1.0.
The file `mm_malloc.h` is taken from upstream `pmm_malloc.h`.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
Add compatibility with gcc <= 12

Implementations based off headers in GCC release 12.3.0.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
Add compatibility with gcc <= 11

Implementations based off headers in GCC release 11.4.0.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
Add compatibility with clang <= 12

Implementations based off headers in LLVM 12.0.0.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
Add compatibility with clang <= 11

Implementations based off headers in LLVM 11.0.0.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

mariasfiraiala reviewed (approved):
Works well, thanks @andreittr! Tested the changes with gcc-11 and clang-14 and it performed just fine.

Reviewed-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

andreittr committed:
Update README & CONTRIBUTING

Reformat to Markdown.
CONTRIBUTING: remove obsolete references, point users at website
README: add a short description

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

RaduNichita commented on Config.uk:
```diff
@@ -1,25 +1,3 @@
-menuconfig LIBINTEL_INTRINSICS
-    bool "Intel Intrinsics - C style functions that provide access Intel instructions"
+config LIBINTEL_INTRINSICS
+    bool "Intel Intrinsics - C style functions that provide access to x86 instructions"
```
```suggestion
    bool "Intel Intrinsics: C style functions that provide access to x86 instructions"
```

RaduNichita replied:
I also think the string before : should start with a lowercase letter, for example [musl](https://github.com/unikraft/lib-musl/blob/staging/Config.uk#L3).

andreittr replied:
Re: `-` vs `:`, there's not really any consistency between libs; musl and builtins are `:`, while libcxx*, libunwind, compiler-rt, newlib, bzip, lwip, and others are `-`. I'm not in favour of picking a particular side here unless we agree for a style project-wide, and until then I err on the side of least change.

Re: capitalization, I'd rather use the project's preferred style, caps & all, since that will be the most familiar for a user, e.g., `musl`, `Python`, `Intel Intrinsics`; they [spell it like this too](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#) (modulo the registered trademark).

RaduNichita reviewed (commented):
I tried running samples from [here](https://github.com/Triple-Z/AVX-AVX2-Example-Code), but for some reason, they don't compile for Unikraft, the issue being

```console
cc1: error: attribute ‘mwait’ argument ‘target’ is unknown` . 
```

Do you know what may be the issue for that?

RaduNichita commented on README.md:
```diff
@@ -1,5 +1,7 @@
+A list of these intrinsics can be browsed at [Intel Intrinsics](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html).
+Normally you would receive these headers together with your compiler, however, due to Unikraft's isolated build environment, we provide them explicitly.
+
+Please refer to `README.md` in the main unikraft repository as well as the [Unikraft Homepage](https://unikraft.org/) for further information.
```
```suggestion
Please refer to `README.md` in the main unikraft repository as well as the [Unikraft Homepage](https://unikraft.org/) for further information.
```

Please provide a link to the `README.md` file

mariasfiraiala replied:
I don't think that's necessary @RaduNichita. Check the other libs, neither of them links the core Unikraft README which is also easily available. Consistency between repos is key.

RaduNichita replied:
OK, fair point here

mariasfiraiala commented:
@RaduNichita @andreittr I'd leave the discussion regarding capitalization and so on for a later time. If the functionality is fine, I'd say we should aim to merge this as part of the 0.14 release.

RaduNichita reviewed (approved):
OK, so  I think we can approve this as it is (as it is necessarily for the Numpy Library) and work in the future in integrating those samples. 

I will open an issue on that after this PR is merged.

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
